pkg/cni/plugin.go

```diff
@@ -0,0 +1,248 @@
+		}
+		return nil
+	})
+	return err
+}
+
+func updateMacvtapIface(macvtapLink netlink.Link, macvtapIface *current.Interface, ifaceName string, netns ns.NetNS) error {
+	err := netns.Do(func(_ ns.NetNS) error {
+		err := ip.RenameLink(macvtapLink.Attrs().Name, ifaceName)
+		if err != nil {
+			_ = netlink.LinkDel(macvtapLink)
+			return fmt.Errorf("failed to rename macvlan to %q: %v", ifaceName, err)
+		}
+
+		updatedLink := macvtapLink
+		updatedLink.Attrs().Name = ifaceName
+		if err := netlink.LinkSetUp(updatedLink); err != nil {
+			return fmt.Errorf("failed to set macvtap iface up: %v", err)
+		}
+		// Re-fetch macvlan to get all properties/attributes
+		contMacvlan, err := netlink.LinkByName(ifaceName)
+		if err != nil {
+			return fmt.Errorf("failed to refetch macvlan %q: %v", ifaceName, err)
+		}
+		macvtapIface.Mac = contMacvlan.Attrs().HardwareAddr.String()
+		macvtapIface.Sandbox = netns.Path()
+
+		return nil
+	})
+	return err
+}
+
+func configureMacvtap(conf *NetConf, ifName string, netns ns.NetNS) (*current.Interface, error) {
+	iface, err := netlink.LinkByName(conf.DeviceID)
+	if err != nil {
+		return nil, fmt.Errorf("failed to lookup device %q: %v", conf.DeviceID, err)
+	}
+	if err := netlink.LinkSetNsFd(iface, int(netns.Fd())); err != nil {
+		return nil, fmt.Errorf("failed to move iface %s to the netns %d because: %v", iface, netns.Fd(), err)
+	}
+	err = netns.Do(func(_ ns.NetNS) error {
+		if err := netlink.LinkSetMTU(iface, conf.MTU); err != nil {
+			return fmt.Errorf("failed to set the macvtap MTU for %s: %v", conf.DeviceID, err)
+		}
+		return nil
+	})
+	macvtap := &current.Interface{Name: ifName}
+	err = configureArp(iface, netns)
+	if err != nil {
+		return nil, err
+	}
+	err = updateMacvtapIface(iface, macvtap, ifName, netns)
+	if err != nil {
+		return nil, err
+	}
+	return macvtap, err
+}
+
+// CmdAdd - CNI interface
+func CmdAdd(args *skel.CmdArgs) error {
+	n, cniVersion, err := loadConf(args.StdinData)
+	if err != nil {
+		return err
+	}
+
+	netns, err := ns.GetNS(args.Netns)
+	if err != nil {
+		return fmt.Errorf("failed to open netns %q: %v", netns, err)
+	}
+	defer netns.Close()
+
+	macvtapInterface, err := configureMacvtap(n, args.IfName, netns)
+	if err != nil {
+		return err
+	}
+
+	// Delete link if err to avoid link leak in this ns
+	defer func() {
+		if err != nil {
+			netns.Do(func(_ ns.NetNS) error {
+				return ip.DelLinkByName(args.IfName)
+			})
+		}
+	}()
+
+	envArgs, err := getEnvArgs(args.Args)
+	if err != nil {
+		return err
+	}
+
+	var mac net.HardwareAddr
+	if envArgs.MAC != "" {
+		mac, err = net.ParseMAC(string(envArgs.MAC))
+		if err != nil {
+			return err
+		}
+	}
+
+	if mac.String() != "" {
+		err = netns.Do(func(_ ns.NetNS) error {
+			macIf, err := netlink.LinkByName(args.IfName)
+			if err != nil {
+				return fmt.Errorf("failed to lookup new macvtapdevice %q: %v", args.IfName, err)
+			}
+
+			if err = netlink.LinkSetHardwareAddr(macIf, mac); err != nil {
```

Comment:
If we set the iface to UP, you need to bring it down in order to change the MAC

Reply:
Moved the code around to make this more explicit (all configuration is performed w/ the interface down, and I set it UP in the `CmdAdd` method)

Please 'resolve' / comment.